openSearchLane.go

```diff
@@ -0,0 +1,277 @@
+
+	ll.clone(&osl.logLane)
+	osl.logLane.writer = log.New(&osl, "", 0)
+	osl.logLane.logFlags = log.Flags() &^ (log.Ldate | log.Ltime)
+
+	l = &osl
+
+	return
+
+}
+
+func (osl *openSearchLane) Connect(openSearchUrl, openSearchPort, openSearchUser, openSearchPass, openSearchIndex, openSearchAppName string) (err error) {
+	if osl.openSearchConnection.client != nil {
+		//close method
+	}
+
+	client, err := newOpenSearchClient(openSearchUrl, openSearchPort, openSearchUser, openSearchPass)
+	if err != nil {
+		return
+	}
+
+	osl.openSearchConnection.client = client
+	osl.openSearchConnection.index = openSearchIndex
+	osl.openSearchConnection.appName = openSearchAppName
+	go osl.flushLogBuffer()
+
+	return
+
+}
+
+func (osl *openSearchLane) Derive() Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(osl.Context, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl2.wg.Add(1)
+	atomic.AddInt32(&osl.openSearchConnection.refCount, 1)
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func (osl *openSearchLane) DeriveReplaceContext(ctx context.Context) Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(ctx, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl2.wg.Add(1)
+	atomic.AddInt32(&osl.openSearchConnection.refCount, 1)
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func newOpenSearchClient(openSearchUrl, openSearchPort, openSearchUser, openSearchPass string) (client *opensearchapi.Client, err error) {
+	client, err = opensearchapi.NewClient(
+		opensearchapi.Config{
+			Client: opensearch.Config{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+				},
+				Addresses: []string{fmt.Sprintf("%s:%s", openSearchUrl, openSearchPort)},
+				Username:  openSearchUser,
+				Password:  openSearchPass,
+			},
+		},
+	)
+
+	return
+}
+
+func (osl *openSearchLane) Write(p []byte) (n int, err error) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	logEntry := string(p)
+
+	fmt.Println(logEntry)
+
+	parentLaneId, ok := osl.Context.Value(parent_lane_id).(string)
+	if !ok {
+		parentLaneId = ""
+	}
+
+	logData := &LogTemplate{
+		AppName:      osl.openSearchConnection.appName,
+		ParentLaneId: parentLaneId,
+		JourneyID:    osl.journeyId,
+		LaneID:       osl.LaneId(),
+		LogMessage:   logEntry,
+	}
+
+	osl.openSearchConnection.logBuffer = append(osl.openSearchConnection.logBuffer, logData)
+
+	return len(p), nil
+}
+
+func (osl *openSearchLane) flushLogBuffer() {
+	backoffDuration := 10 * time.Second
+	ticker := time.NewTicker(backoffDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			osl.openSearchConnection.mu.Lock()
```

Comment:
This should take possession of logBuffer. Assuming function context is `func (osc *openSearchConnection)` then

```
osc.mu.Lock()
lb := osc.logBuffer
osc.logBuffer = make([]*LogTemplate, 0, someReasonableSize)
osc.mu.Unlock()

// now caller is free to log more, not blocked by this I/O

// rest of the code sends the bulkInsert of the owned
```